### ISSUE TITLE
javabridge and PyQT

### ISSUE THREAD

**Issue by Heerpa** [LeeKamentsky/python-javabridge#97]
Not sure this is the right way and place to ask this, but it kind of seems not completely out of place. This is probably not a bug in the package but rather a possible impulse for another example in the documentation.
I am facing issues caused by the fact that javabridge can open a jvm only once. No examples or issues I read seemed to apply.
Basically, I am using PyQT to make a gui that is supposed to load images using bioformats on user input (can be multiple times). It works once if I start the jvm within the method using bioformats, but not if the jvm is started in **init** of the GUI window, nor in the main function. I guess this is something to do with me starting it in the wrong thread, as discussed in [http://pythonhosted.org/javabridge/start_kill.html], but don't see how this is meant to be done. I'd appreciate any suggestions.
Thanks,
   Heinrich
Here is a minimal working example:

`

```
from PyQt4 import QtGui
import bioformats
import javabridge
import sys
import traceback


class Window(QtGui.QWidget):
    '''the GUI window class'''
    def __init__(self):
        super().__init__()
        # javabridge.start_vm(class_path=bioformats.JARS)
        self.setWindowTitle('Testing QtGui and Bioformats')
        vbox = QtGui.QVBoxLayout()
        self.setLayout(vbox)

        self.pathedit = QtGui.QTextEdit('./TestData/3chan.tif')
        vbox.addWidget(self.pathedit)
        but_load = QtGui.QPushButton('load image')
        but_load.clicked.connect(self.load)
        vbox.addWidget(but_load)
        self.lbl_info = QtGui.QLabel('')
        vbox.addWidget(self.lbl_info)
        vbox.addStretch(1)
        but_test = QtGui.QPushButton('Do something with the image')
        but_test.clicked.connect(self.do)
        vbox.addWidget(but_test)

    def __del__(self):
        # javabridge.kill_vm()
        pass

    def load(self):
        path = self.pathedit.toPlainText()
        print('loading the image ', path)
        javabridge.start_vm(class_path=bioformats.JARS)
        meta = bioformats.get_omexml_metadata(path)
        o = bioformats.OMEXML(meta)
        print('Name:', o.image(0).Name)
        print('size x:', o.image(0).Pixels.SizeX)
        self.lbl_info.setText('Name: ' + o.image(0).Name +
                              '\tsize x: ' + str(o.image(0).Pixels.SizeX))
        javabridge.kill_vm()

    def do(self):
        print('doing something with the image')


if __name__ == '__main__':
    # start java virtual machine for loading images using bioformats
    # javabridge.start_vm(class_path=bioformats.JARS)

    try:
        app = QtGui.QApplication(sys.argv)
        window = Window()
        window.show()
    finally:
        # again, for bioformats
        # javabridge.kill_vm()
        pass

    def excepthook(type, value, tback):
        message = ''.join(traceback.format_exception(type, value, tback))
        errorbox = QtGui.QMessageBox.critical(window, 'An error occured',
                                              message)
        errorbox.exec_()
        sys.__excepthook__(type, value, tback)
    sys.excepthook = excepthook

    sys.exit(app.exec_())
```

`

**Comment by LeeKamentsky** [comment]
It's a limitation of the JVM and it really is a pain to manage in just about every substantial program that uses the Javabridge. There was a recent PR (#71) that helped a little so you can write:

    import javabridge
    with javabridge.vm(class_path=bioformats.JARS):
        app = QtGui.QApplication(sys.argv)
        window = Window()
        window.show()

if that helps at all. In any case, believe me, I feel your pain ;-)

**Comment by Heerpa** [comment]
Thanks very much! I inserted the javasrcipt "hello world" after with, and "hello window" in __init__, and "hello load" in the beginning of load(self). The first two work, but load does not. As using bioformats.
Just before giving up, I noticed your comment about javabridge.attach() and javabridge.detach() in said PR (#71). Flanking the load method by these helped and made the program work!

For later reference, here is the working code:


    from PyQt4 import QtGui
    import bioformats
    import javabridge
    import sys
    import traceback


    class Window(QtGui.QWidget):
        '''the GUI window class'''
        def __init__(self):
            print(javabridge.run_script('java.lang.String.format("Hello, %s!", greetee);',dict(greetee='window')))
            super().__init__()
            self.setWindowTitle('Testing QtGui and Bioformats')
            vbox = QtGui.QVBoxLayout()
            self.setLayout(vbox)

            self.pathedit = QtGui.QTextEdit('./TestData/3chan.tif')
            vbox.addWidget(self.pathedit)
            but_load = QtGui.QPushButton('load image')
            but_load.clicked.connect(self.load)
            vbox.addWidget(but_load)
            self.lbl_info = QtGui.QLabel('')
            vbox.addWidget(self.lbl_info)
            vbox.addStretch(1)
            but_test = QtGui.QPushButton('Do something with the image')
            but_test.clicked.connect(self.do)
            vbox.addWidget(but_test)

        def __del__(self):
            pass

        def load(self):
            javabridge.attach()
            print(javabridge.run_script('java.lang.String.format("Hello, %s!", greetee);',dict(greetee='load')))
            path = self.pathedit.toPlainText()
            print('loading the image ', path)
            meta = bioformats.get_omexml_metadata(path)
            o = bioformats.OMEXML(meta)
            print('Name:', o.image(0).Name)
            print('size x:', o.image(0).Pixels.SizeX)
            self.lbl_info.setText('Name: ' + o.image(0).Name +
                                  '\tsize x: ' + str(o.image(0).Pixels.SizeX))
            javabridge.detach()

        def do(self):
            print('doing something with the image')


    if __name__ == '__main__':
        
        try:
            # start java virtual machine for loading images using bioformats
            with javabridge.vm(class_path=bioformats.JARS):
                print(javabridge.run_script('java.lang.String.format("Hello, %s!", greetee);',dict(greetee='world')))
                app = QtGui.QApplication(sys.argv)
                window = Window()
                window.show()
        finally:
            # again, for bioformats
            javabridge.kill_vm()
            pass

        def excepthook(type, value, tback):
            message = ''.join(traceback.format_exception(type, value, tback))
            errorbox = QtGui.QMessageBox.critical(window, 'An error occured',
                                                  message)
            errorbox.exec_()
            sys.__excepthook__(type, value, tback)
        sys.excepthook = excepthook

        sys.exit(app.exec_())

**Comment by Heerpa** [comment]
Unfortunately, there have been issues with attach()ing again.
All of the above still works on my Mac OS X (now Sierra, back when this was written Yosemite).
I now want this code to run on a Windows (Win Server 2012 R2 Standard) machine. Here, python crashes in load(self) at 
`javabridge.attach()`
I do get the earlier outputs
```
Hello World
Hello Window

```
In a test case where I open an image using python-bioformats and javabridge without PyQt, everything works fine.
The issue must lie in a different way of opening threads in Windows and OS X. 
I'd be grateful for any pointers.

**Comment by Heerpa** [comment]
I now looked deeper into it and realized that the issue comes up when a callback is called (via xx.clicked.connect), not when the QtGui window class is initialized. This makes sense as then, a new thread is opened.
Digging deeper into javabridge, the crash happens somewhere in JB_VM.attach_as_daemon()
https://github.com/LeeKamentsky/python-javabridge/blob/3e43fcec733e86b7b90d6a75f4ef3cfa44a2a3c7/_javabridge.pyx#L716

what I get is a crash report "Python has stopped working" with the additional information:
Problem signature:
  Problem Event Name:	APPCRASH
  Application Name:	python.exe
  Application Version:	3.5.2150.1013
  Application Timestamp:	577be340
  Fault Module Name:	_javabridge.cp35-win_amd64.pyd
  Fault Module Version:	0.0.0.0
  Fault Module Timestamp:	58ade266
  Exception Code:	c0000005
  Exception Offset:	000000000001ea79
  OS Version:	6.3.9600.2.0.0.16.7
  Locale ID:	1033
  Additional Information 1:	f71f
  Additional Information 2:	f71feae0df1f53ff0658441f16bbf537
  Additional Information 3:	2d61
  Additional Information 4:	2d611f9bd02a8bd14272407ba02e92dc

Read our privacy statement online:
  http://go.microsoft.com/fwlink/?linkid=280262

If the online privacy statement is not available, please read our privacy statement offline:
  C:\Windows\system32\en-US\erofflps.txt

**Comment by Heerpa** [comment]
Apparently, the
`with javabridge.vm(class_path=bioformats.JARS):`
statement from #71 kills the JVM directly after showing the window (__init__ still runs). When the 'load' button is clicked, __vm from _javabridge still exists, but its __vm.vm pointer is NULL. This makes sense as jutil.vm__exit__() is called after window init.
This is true when I tested it under Windows, while under OSX, jutil.vm.__exit__() is called only when the wiindows are closed.

my suggestion would be to add
assert __vm.is_active()
before line
https://github.com/LeeKamentsky/python-javabridge/blob/master/_javabridge.pyx#L406
to make this more obvious.

For the minimal working example, the code running on OSX as well as Windows now looks like this:
```

from PyQt4 import QtGui
import bioformats
import javabridge
import sys
import traceback


class Window(QtGui.QWidget):
    '''the GUI window class'''
    def __init__(self):
        # javabridge.attach()  # no need to attach as this is in the main thread.
        print(javabridge.run_script('java.lang.String.format("Hello, %s!", greetee);',dict(greetee='window')))
        # javabridge.detach()
        super().__init__()
        self.setWindowTitle('Testing QtGui and Bioformats')
        vbox = QtGui.QVBoxLayout()
        self.setLayout(vbox)

        self.pathedit = QtGui.QTextEdit('./TestData/3chan.tif')
        vbox.addWidget(self.pathedit)
        but_load = QtGui.QPushButton('load image')
        but_load.clicked.connect(self.load)
        vbox.addWidget(but_load)
        self.lbl_info = QtGui.QLabel('')
        vbox.addWidget(self.lbl_info)
        vbox.addStretch(1)
        but_test = QtGui.QPushButton('Do something with the image')
        but_test.clicked.connect(self.do)
        vbox.addWidget(but_test)
        # javabridge.attach()
        print(javabridge.run_script('java.lang.String.format("Hello, %s!", greetee);',dict(greetee='window is here')))
        # javabridge.detach()

    def closeEvent(self, event):
        print('closing')
        javabridge.attach()
        print(javabridge.run_script('java.lang.String.format("Bye, %s!", greetee);',dict(greetee='window')))
        javabridge.detach()
        javabridge.kill_vm()
        event.accept()

    def load(self):
        javabridge.attach()
        print(javabridge.run_script('java.lang.String.format("Hello, %s!", greetee);',dict(greetee='load')))
        javabridge.detach()
        self.env_ = None
        path = self.pathedit.toPlainText()
        print('loading the image ', path)
        meta = bioformats.get_omexml_metadata(path)
        o = bioformats.OMEXML(meta)
        print('Name:', o.image(0).Name)
        print('size x:', o.image(0).Pixels.SizeX)
        self.lbl_info.setText('Name: ' + o.image(0).Name +
                              '\tsize x: ' + str(o.image(0).Pixels.SizeX))

    def do(self):
        print('doing something with the image')


if __name__ == '__main__':

    try:
        # start java virtual machine for loading images using bioformats
        # with javabridge.vm(class_path=bioformats.JARS):
        #     # javabridge.attach()
        #     print(javabridge.run_script('java.lang.String.format("Hello, %s!", greetee);',dict(greetee='world')))
        #     # javabridge.detach()
        #     app = QtGui.QApplication(sys.argv)
        #     window = Window()
        #     window.show()
        javabridge.start_vm(class_path=bioformats.JARS)
        print(javabridge.run_script('java.lang.String.format("Hello, %s!", greetee);',dict(greetee='world')))
        app = QtGui.QApplication(sys.argv)
        window = Window()
        window.show()
    finally:
        pass

    def excepthook(type, value, tback):
        message = ''.join(traceback.format_exception(type, value, tback))
        errorbox = QtGui.QMessageBox.critical(window, 'An error occured',
                                              message)
        errorbox.exec_()
        sys.__excepthook__(type, value, tback)
    sys.excepthook = excepthook

    sys.exit(app.exec_())

```

**Comment by LeeKamentsky** [comment]
@heerpa, I added the assertion you requested. Also, thank you for posting your examples - they're a welcome addition to the documentation.
